web_console/rpc/server.js

```diff
@@ -143,20 +143,116 @@ async function deleteJob(call, callback) {
       },
     });
     if (!job) throw new Error('Job not found');
-    await k8s.deleteFLApp(NAMESPACE, job.name);
+    if (data.status == 'started') {
+      await k8s.deleteFLApp(NAMESPACE, job.name);
+    }
     await job.destroy({ force: true });
     callback(null, { message: 'Delete job successfully' });
   } catch (err) {
     callback(err);
   }
 }
 
+async function updateJob(call, callback) {
+  try {
+    const federation = authenticate(call.metadata);
+
+    const {
+      name,
+      job_type,
+      client_ticket_name: server_ticket_name,
+      server_ticket_name: client_ticket_name,
+      server_params,
+      status,
+    } = call.request;
+
+    const new_job = {
+      name, job_type, client_ticket_name, server_ticket_name,
+      server_params, status,
+    }
+
+    const old_job = await Job.findOne({
+      where: {
+        name: { [Op.eq]: name },
+      },
+    });
+    if (!old_job) throw new Error('Job not found');
+
+    if (old_job.status === 'error') {
+      throw new Error("Cannot update errored job");
+    }
+    if (old_job.status === 'started' && new_job.status != 'stopped') {
+      throw new Error("Cannot change running job");
+    }
+    if (job_type != old_job.job_type) {
+      throw new Error("Cannot change job type");
+    }
+
+    const ticketRecord = await Ticket.findOne({
+      where: {
+        name: { [Op.eq]: client_ticket_name },
+      },
+    });
+    if (!ticketRecord) throw new Error('Ticket not found');
+
+    if (ticketRecord.federation_id != old_job.federation_id) {
+      throw new Error("Cannot change job federation");
+    }
+
+    const params = JSON.parse(server_params);
+    validateTicket(ticketRecord, params);
+
+    if (old_job.status === 'started' && status === 'stopped') {
+      flapp = (await k8s.getFLApp(NAMESPACE, name)).flapp;
+      pods = (await k8s.getFLAppPods(NAMESPACE, name)).pods;
+      old_job.k8s_meta_snapshot = JSON.stringify({flapp, pods});
+      await k8s.deleteFLApp(NAMESPACE, name);
+    } else if (old_job.status === 'stopped' && new_job.status === 'started') {
+      const args = serverGenerateYaml(federation, new_job, ticketRecord);
+      await k8s.createFLApp(NAMESPACE, args);
+    }
+
+    old_job.client_ticket_name = new_job.client_ticket_name;
+    old_job.server_ticket_name = new_job.server_ticket_name;
+    old_job.server_params = new_job.server_params;
+    old_job.status = new_job.status;
+    const data = await old_job.save();
+
+    callback(null, {
+      data: {
+        name: data.name,
+        job_type: data.job_type,
+        client_ticket_name: data.server_ticket_name,
+        server_ticket_name: data.client_ticket_name,
+        server_params: JSON.stringify(data.server_params),
+      },
+      status: data.status,
+    });
+  } catch (err) {
+    callback(err);
+  }
+}
+
+/**
+ * get available tickets of current federation
+ */
+async function heartBeat(call, callback) {
```

Comment:
it'd be better to respond directly for heartbeat detection, any other code is overkill :)